raster/r.patch/main.c

```diff
@@ -130,46 +156,109 @@ int main(int argc, char *argv[])
     rname = opt2->answer;
     outfd = Rast_open_new(new_name = rname, out_type);
 
-    presult = Rast_allocate_buf(out_type);
-    patch = Rast_allocate_buf(out_type);
+    presult = G_malloc(nprocs * sizeof *presult);
+    patch = G_malloc(nprocs * sizeof *patch);
+    for (t = 0; t < nprocs; t++) {
+        presult[t] = Rast_allocate_buf(out_type);
+    }
+    for (t = 0; t < nprocs; t++) {
+        patch[t] = Rast_allocate_buf(out_type);
+    }
 
     Rast_get_window(&window);
     nrows = Rast_window_rows();
     ncols = Rast_window_cols();
 
+    bufrows = atoi(memory->answer) * (((1 << 20) / sizeof(DCELL)) / ncols);
+    /* set the output buffer rows to be at most covering the entire map */
+    if (bufrows > nrows) {
+        bufrows = nrows;
+    }
+    /* but at least the number of threads */
+    if (bufrows < nprocs) {
+        bufrows = nprocs;
+    }
+
+    outbuf = G_malloc(out_cell_size * ncols * bufrows);
+
     G_verbose_message(_("Percent complete..."));
-    for (row = 0; row < nrows; row++) {
-        double north_edge, south_edge;
-
-        G_percent(row, nrows, 2);
-        Rast_get_row(infd[0], presult, row, out_type);
-
-        north_edge = Rast_row_to_northing(row, &window);
-        south_edge = north_edge - window.ns_res;
-
-        if (out_type == CELL_TYPE)
-            Rast_update_cell_stats((CELL *) presult, ncols, &statf[0]);
-        for (i = 1; i < nfiles; i++) {
-            /* check if raster i overlaps with the current row */
-            if (south_edge >= cellhd[i].north ||
-                north_edge <= cellhd[i].south ||
-                window.west >= cellhd[i].east ||
-                window.east <= cellhd[i].west)
-                continue;
-
-            Rast_get_row(infd[i], patch, row, out_type);
-            if (!do_patch(presult, patch, &statf[i], ncols, out_type,
-                          out_cell_size, use_zero))
-                break;
+
+    int computed = 0;
+    int written = 0;
+
```

Comment:
`row = 0` to avoid
```
main.c:250:5: warning: ‘row’ may be used uninitialized in this function [-Wmaybe-uninitialized]
     G_percent(row, nrows, 2);
```

Reply:
Yeap fixed it! Supposed to be `G_percent(nrows, nrows, 2)`.